src/re/multi_grid/kernel.py

```diff
@@ -0,0 +1,312 @@
+            unique = jnp.full((buffer_size,) + shp, jnp.nan)
+
+            (unique, inv), invid = scan(
+                partial(scanned_amend_unique, shift=shift), (unique, inv), indices
+            )
+            _, idx = np.unique(invid, return_index=True)
+            n = idx.size
+            if n >= unique.shape[0] or not np.all(np.isnan(unique[n:])):
+                raise ValueError("`mat_buffer_size` too small")
+            uids = indices[idx]
+            uids = gridf_at.flatindex2index(uids[np.newaxis, :])
+            uindices.append(uids)
+            invindices.append(inv)
+            indexmaps.append(_IdxMap(shift, gridf_at.index2flatindex))
+
+        return self.replace(
+            _cim=_CompressedIndexMap(
+                base_kernel=None,
+                kernels=None,
+                uindices=uindices,
+                indexmaps=indexmaps,
+                invindices=invindices,
+            )
+        )
+
+    def compress_matrices(self):
+        """Recompute the kernel matrices while keeping the tables for speedy lookups
+        fixed.
+
+        This is useful for updating the covariance function of a kernel without
+        recreating the tables in an expensive optimization.
+        """
+        assert self._cim is not None
+
+        base_kernel = self.compute_matrices(jnp.array([-1]), None)
+        kernels = tuple(
+            self.compute_matrices(ii, ll) for ll, ii in enumerate(self._cim.uindices)
+        )
+        cim = self._cim._replace(base_kernel=base_kernel, kernels=kernels)
+        return self.replace(_cim=cim)
+
+    def compress(self, *args, **kwargs):
+        """Optimize the kernel matrices for the given grid.
+
+        This links kernel matrices by similarity and subsequently looks them up via
+        tables. Compressing the kernel in this way has the potential to drastically
+        reduce the memory requirements of applying it and make the inference of
+        kernels faster.
+        """
+        return self.compress_indices(*args, **kwargs).compress_matrices()
+
+
+def _suitable_window_size(grid_at_level, default=3) -> Tuple[int]:
+    wsz = []
+    for g in grid_at_level.raw_grids:
+        if isinstance(g, HEALPixGridAtLevel):  # special-case HEALPix
+            wsz += [9]
+        elif isinstance(g, OpenGridAtLevel):
+            wsz += list(g.padding * 2 + 1)
+        else:
+            wsz += [default] * g.ndim
+    return tuple(wsz)
+
+
+class ICRKernel(Kernel):
+    """Full ICR implementation taking an arbitrary grid and a covariance function."""
+
+    def __init__(self, grid, covariance, *, window_size=None, _cim=None):
+        self._covariance_elem = covariance
+        if window_size is None:
+            window_size = tuple(
+                _suitable_window_size(grid.at(lvl)) for lvl in range(grid.depth)
+            )
+        self._window_size = tuple(
+            np.broadcast_to(window_size, (grid.depth,) + grid.at(0).shape.shape)
```

Comment:
isn't this just `(grid.deph, grid.at(0).ndim)`?

Reply:
fair point 😅 fixed :smiley: